### PR TITLE
Deprecate old Reg with nulls constructor

### PR DESCRIPTION
**Pull request by ducky64**
*cleaning up Chisel, removing null at a time*

Implementation for resolution in #438. With some clean-up in Wire, eliminating more nulls.

Up for debate is a Reg constructor with an explicit model and init, which isn't supported under RegNext/RegInit. I just did Reg.Init for now, but feel free to come up with a better API. Conceptually, Reg creates a type with explicit model, and its Init variant allows an optional initializer. Specifying init after the Reg is declared would not be easy to implement.

Also for consideration is Reg semantics. For compatibility reasons, if next and init are defined (here: RegNext(next, init)), the resulting model type has no width, but is init is defined without next, the resulting mode type has width in init, but only if it is explicitly specified. The rule is a mouthful, and the next no width prioritization seems nonsensical. I'd advocate for init-derived width to take priority. And perhaps a better (internal) API for getting model from an init, since that is also re-used in Wire.

### CODE REVIEW

**Comment by aswaterman** [comment]
Deriving the width from the max of init and next seems less surprising.  (In practice, this is what the current no-width rule accomplishes.)

**Comment by ducky64** [comment]
I like that width from max of init and next is consistent with the cloneSupertype semantics from #450, but one issue might be:
```scala
val reg = RegInit(0.U)
reg := reg + 1
```
But I think that will produce a broken result now and is (in any case) actually underconstrained. This is slightly more insidious:
```scala
val myWire = ...  // something that ends up as 4 bits
val reg = RegInit(0.U)  // 0 or 1 bits, don't remember
reg := myWire  // bits are dropped
```
However, in the above, `RegNext(myWire, 0.U)` should produce correct results.

**Comment by aswaterman** [comment]
What about no-width for all cases that ```t``` is not given (or ```t``` has no width)?

**Comment by ducky64** [comment]
Yeah, those are those cases. `RegInit` (different from `Reg.Init`) only takes a `init` value (as it did before) and automatically derives `t` based on that value. Currently, there's a clear width operation involved (unless init was literal with explicit width).

For `RegNext`: width is always cleared (regardless of value of `init` or whether it is present).

For `Reg(t)`: `t` is used directly (width is preserved)

**Comment by jackkoenig** [comment]
Can you resolve the conflicts?

**Comment by ducky64** [comment]
Updated.

We might want to do something about the awful logic that RegInit/RegNext/Wire uses to select the data model of the output type, possibly using cloneSupertype from the other PR. In particular, all the RegNext constructs should just use next.(chisel)CloneType. A debate is possible for what happens in the RegInit case and Wire (with init) case.

File: coreMacros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoTransform.scala [hunk 1]
```diff
@@ -27,14 +27,6 @@ trait SourceInfoTransformMacro {
   def implicitCompileOptions = q"implicitly[_root_.chisel3.core.CompileOptions]"
 }
 
-class WireTransform(val c: Context) extends SourceInfoTransformMacro {
-  import c.universe._
-  def apply[T: c.WeakTypeTag](t: c.Tree): c.Tree = {
-    val tpe = weakTypeOf[T]
-    q"$thisObj.do_apply($t, null.asInstanceOf[$tpe])($implicitSourceInfo, $implicitCompileOptions)"
-  }
-}
```
Comment (jackkoenig) [inline coreMacros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoTransform.scala]: Why are these going away? Can we not get source locators with Reg and Wire?

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 2]
```diff
@@ -334,28 +334,37 @@ abstract class Data extends HasId {
...
+    val model = (init.litArg match {
+      // For e.g. Wire(init=UInt(0, k)), fix the Reg's width to k
+      case Some(lit) if lit.forcedWidth => init.chiselCloneType
+      case _ => init match {
+        case init: Bits => init.cloneTypeWidth(Width())
+        case init => init.chiselCloneType
+      }
+    }).asInstanceOf[T]
+    apply(model, init)
+  }
```
Comment (jackkoenig) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Should we move to `WireInit` as well? I guess this can be a future PR

File: src/test/scala/chiselTests/Reg.scala [hunk 3]
```diff
@@ -11,39 +11,32 @@ import chisel3.core.DataMirror
...
     class RegForcedWidthTester extends BasicTester {
-      val reg2 = Reg(init=20.U(7.W))
-      reg2.getWidth should be (7)
+      val reg1 = RegInit(20.U)
+      DataMirror.widthOf(reg1).known should be (false)
+      val reg2 = RegInit(20.U(7.W))
+      DataMirror.widthOf(reg2) should be (7.W)
     }
     elaborate{ new RegForcedWidthTester }
   }
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/Reg.scala]: Larger testing point. We really should have tests that *specify* and *enforce* how width inference should work. For example:
```scala
  "Wire" should "infer the right widths" in {
    assertTesterFails(new BasicTester {
      val w1 = Wire(init = 0.U)
      chisel3.assert(~w1 === 0xf.U)
      stop()
    })
    assertTesterPasses(new BasicTester {
      val w1 = Wire(init = 1.U)
      w1 := 0.U(4.W)
      chisel3.assert(~w1 === 0xf.U)
      stop()
    })
    assertTesterFails(new BasicTester {
      val w1 = Wire(init = 1.U)
      w1 := 0.U(5.W)
      chisel3.assert(~w1 === 0xf.U)
      stop()
    })
  }
```
Tests like these that check the width of the final Verilog. We could even have a helper function like `assertInferredWidth` or something to make those assertions a little more clear. We need this for all of the operators as well.

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala [hunk 4]
```diff
@@ -2,87 +2,93 @@
...
-          case init: Bits => init.cloneTypeWidth(Width())
-          case _ => init.chiselCloneType
-        }).asInstanceOf[T]
-      }
-    } else {
-      throwException("cannot infer type")
+  /** Creates a register with reset value. Internal API, users should use RegInit instead.
+    * TODO: should this be a public API, where there is a specified init of different type than the reg?
+    */
+  def Init[T <: Data](t: T, init: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
```
Comment (jackkoenig) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: Why `Init` instead of just another `apply` method?

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala [hunk 5]
```diff
@@ -2,87 +2,93 @@
...
-      if (compileOptions.declaredTypeMustBeUnbound) {
-        Binding.checkUnbound(t, s"t ($t) must be unbound Type. Try using cloneType?")
-      }
-      t.chiselCloneType
-    } else if (next ne null) {
-      (next match {
-        case next: Bits => next.cloneTypeWidth(Width())
-        case _ => next.chiselCloneType
-      }).asInstanceOf[T]
+trait RegFactory {
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: Ideally no, but it needs to be public so Chisel (compatibility package) can extend it...

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala [hunk 6]
```diff
@@ -2,87 +2,93 @@
...
-        case Some(lit) if lit.forcedWidth => init.chiselCloneType
-        case _ => (init match {
-          case init: Bits => init.cloneTypeWidth(Width())
-          case _ => init.chiselCloneType
-        }).asInstanceOf[T]
-      }
-    } else {
-      throwException("cannot infer type")
+  /** Creates a register with reset value. Internal API, users should use RegInit instead.
+    * TODO: should this be a public API, where there is a specified init of different type than the reg?
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: You can pass in `t` (model) and `init` of different types

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala (hunk 4 above)
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: I liked them to be all within the same oject. Also, RegInit is taken, and this is a underlying class for the compatibility mode Reg(t, next, init).

File: coreMacros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoTransform.scala (hunk 1 above)
Comment (ducky64) [inline coreMacros/src/main/scala/chisel3/internal/sourceinfo/SourceInfoTransform.scala]: They're unnecessary, the main reason we needed transforms is so you can chain an apply. But if you chain an apply after a Reg / Wire call, you're basically discarding bits, which seems like an odd use case.

File: src/test/scala/chiselTests/Reg.scala (hunk 3 above)
Comment (ducky64) [inline src/test/scala/chiselTests/Reg.scala]: Good point. Though I think we may also want to discuss when the width should be cleared - the current code structure (with litArg, etc in the Wire and Reg constructors) makes the convoluted logic explicit.

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala (hunk 6 above)
Comment (jackkoenig) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: Do you have an example of when that would happen?

**Comment by jackkoenig** [comment]
I'm a bit confused by the width inference I'm seeing:

* Wires
  1. `val w1 = Wire(init = 3.U)`
      `wire w1 : UInt`
  1. `val w2 = Wire(init = 3.U(32.W))`
      `wire w2 : UInt<32>`
  1. `val w3 = Wire(init = w1)`
      `wire w3 : UInt`
  1. `val w4 = Wire(init = w2)`
      `wire w4 : UInt`
* RegNext
  1. `val rn1 = RegNext(3.U)`
      `reg rn1 : UInt, clock`
  1. `val rn2 = RegNext(3.U, 4.U)`
      `reg rn2 : UInt, clock with : (reset => (reset, UInt<3>("h04")))`
  1. `val rn3 = RegNext(3.U, 4.U(32.W))`
      `reg rn3 : UInt, clock with : (reset => (reset, UInt<32>("h04")))`
  1. `val rn4 = RegNext(w1)`
      `reg rn4 : UInt, clock`
  1. `val rn5 = RegNext(w2)`
      `reg rn5 : UInt, clock`
  1. `val rn6 = RegNext(3.U, w1)`
      `reg rn6 : UInt, clock with : (reset => (reset, w1))`
  1. `val rn7 = RegNext(3.U, w2)`
      `reg rn7 : UInt, clock with : (reset => (reset, w2))`
* RegEnable
  1. `val re1 = RegEnable(3.U, true.B)`
      `reg re1 : UInt<2>, clock`
  1. `val re2 = RegEnable(w1, true.B)`
      `reg re2 : UInt, clock`
  1. `val re3 = RegEnable(w2, true.B)`
      `reg re3 : UInt<32>, clock`          

Wires seem to make sense and it seems to match the code. I'm a bit confused by the fact that the `RegNext` initialized by a literal with a width does not get its width set by that. I'm also confused by the fact that `RegEnable` has its width set if the width of it's "next" is known. I can see why in the code, but it doesn't make sense to me as the right API. I feel like `RegEnable` should match `RegNext`

**Comment by jackkoenig** [comment]
We had also (after the PR I believe), discussed `RegNextInit` rather than `RegNext(next, init)`. What are people's thoughts on that?

**Comment by ducky64** [comment]
We might want to discuss the width interference / dropping behavior. I think the way I implemented this PR is to match the previous semantics, but whether they are good or sane can be an open question. The way I structured it should make the logic very clear (and possibly clearly ugly).

**Comment by jackkoenig** [comment]
Yeah I think you made the convoluted logic much more clear than it was which I greatly appreciate. I would like to see a clear explanation of the width inference rules which I'm thinking I'll try to get started here soon on a `WidthInferenceSpec` in the flavor of my examples above.

**Comment by ducky64** [comment]
Resolution:
- Try to make RegFactory and friends private. Chisel (compatibility) should redefine it without extending.
- Will push through this PR preserving current behavior. In the future, we will streamline the width specification rules, with the rule being that widths are cleared (inferred) unless you pass an explicit type (not just an initializer / next)
- Will move Reg.Init constructor to RegInit.apply(type, init), overloaded with RegInit.apply(init)

**Comment by ducky64** [comment]
Above points implemented.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 7]
```diff
@@ -334,28 +334,37 @@ abstract class Data extends HasId {
...
 
 object Wire {
-  def apply[T <: Data](t: T): T = macro WireTransform.apply[T]
-
   // No source info since Scala macros don't yet support named / default arguments.
-  def apply[T <: Data](dummy: Int = 0, init: T)(implicit compileOptions: CompileOptions): T =
-    do_apply(null.asInstanceOf[T], init)(UnlocatableSourceInfo, compileOptions)
+  def apply[T <: Data](dummy: Int = 0, init: T)(implicit compileOptions: CompileOptions): T = {
+    val model = (init.litArg match {
+      // For e.g. Wire(init=UInt(0, k)), fix the Reg's width to k
```
Comment (chick) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Minor nit, comment shows old style UInt literal
Also, I agree with @jackkoenig about API expand to have WireInit

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 8]
```diff
@@ -334,28 +334,37 @@ abstract class Data extends HasId {
...
+        case init => init.chiselCloneType
+      }
+    }).asInstanceOf[T]
+    apply(model, init)
+  }
 
   // No source info since Scala macros don't yet support named / default arguments.
-  def apply[T <: Data](t: T, init: T)(implicit compileOptions: CompileOptions): T =
-    do_apply(t, init)(UnlocatableSourceInfo, compileOptions)
+  def apply[T <: Data](t: T, init: T)(implicit compileOptions: CompileOptions): T = {
```
Comment (chick) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Can't there be some scaladoc comments for these two apply functions

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala [hunk 9]
```diff
@@ -2,88 +2,97 @@
...
-    // option, but introduces cumbersome syntax (wrap everything in a Some()).
-    // Implicit conversions to Option (or similar) types were also considered,
-    // but Scala's type inferencer and implicit insertion isn't smart enough
-    // to resolve all use cases. If the type inferencer / implicit resolution
-    // system improves, this may be changed.
-    val x = makeType(compileOptions, t, next, init)
+object RegInit {
+  /** Creates a register given an explicit type and an initialization (reset) value.
+    */
+  def apply[T <: Data](t: T, init: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
```
Comment (chick) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: Is this for backward compatibility.  If not it seems like just another non-orthogonal method. 
Better to use RegInit(7.U(64.W) than RegInit(UInt(64.W), 7.U).  If it's necessary is there any small value in
making the single argument apply come first and be more visible in the scaladoc

File: src/main/scala/chisel3/compatibility.scala [hunk 10]
```diff
@@ -165,7 +165,51 @@ package object Chisel {     // scalastyle:ignore package.object.name
...
+      * empty to not update unless assigned to using the := operator)
+      * @param init: initialization value on reset (or empty for uninitialized,
+      * where the register value persists across a reset)
+      *
+      * @note this may result in a type error if called from a type parameterized
+      * function, since the Scala compiler isn't smart enough to know that null
+      * is a valid value. In those cases, you can either use the outType only Reg
+      * constructor or pass in `null.asInstanceOf[T]`.
+      */
+    def apply[T <: Data](t: T = null, next: T = null, init: T = null)
```
Comment (chick) [inline src/main/scala/chisel3/compatibility.scala]: This code is an exact copy of the code in the chisel3/package.scala, is the package stuff too convoluted to invoke DRY

File: src/main/scala/chisel3/util/Reg.scala [hunk 11]
```diff
@@ -6,41 +6,21 @@ import chisel3._
...
-  def apply[T <: Data](next: T, init: T): T = Reg[T](null.asInstanceOf[T], next, init)
-}
-
-object RegInit {
-  /** Returns a register pre-initialized (on reset) to the specified value.
-    */
-  def apply[T <: Data](init: T): T = Reg[T](null.asInstanceOf[T], null.asInstanceOf[T], init)
-}
-
 object RegEnable {
```
Comment (chick) [inline src/main/scala/chisel3/util/Reg.scala]: Why did RegEnable stay here when everybody else went to to core

File: src/test/scala/chiselTests/Assert.scala [hunk 12]
```diff
@@ -27,16 +27,16 @@ class SucceedingAssertTester() extends BasicTester {
...
+  val b = RegInit(0xbeef.U)
   assert(a === b)
 }
 
 // This relies on reset being asserted for 3 or more cycles
 class PipelinedResetTester extends BasicTester {
   val module = Module(new PipelinedResetModule)
 
-  module.reset := Reg(next = Reg(next = Reg(next = reset)))
+  module.reset := RegNext(RegNext(RegNext(reset)))
```
Comment (chick) [inline src/test/scala/chiselTests/Assert.scala]: This looks nice, better than old

File: src/test/scala/chiselTests/MultiClockSpec.scala [hunk 13]
```diff
@@ -4,7 +4,7 @@ package chiselTests
 
 import chisel3._
 import chisel3.experimental.{withClockAndReset, withClock, withReset}
-import chisel3.util.{Counter, RegInit}
+import chisel3.util.{Counter}
```
Comment (chick) [inline src/test/scala/chiselTests/MultiClockSpec.scala]: No big deal but style checker complains about braces when only one thing being imported

File: src/test/scala/chiselTests/Reg.scala [hunk 14]
```diff
@@ -11,39 +11,32 @@ import chisel3.core.DataMirror
...
-    }
-  }
-
-  "A Reg" should "be of the same type and width as outType, if specified" in {
+  "Reg" should "be of the same type and width as t" in {
     class RegOutTypeWidthTester extends BasicTester {
-      val reg = Reg(t=UInt(2.W), next=Wire(UInt(3.W)), init=20.U)
-      reg.getWidth should be (2)
+      val reg = Reg(UInt(2.W))
+      DataMirror.widthOf(reg) should be (2.W)
```
Comment (chick) [inline src/test/scala/chiselTests/Reg.scala]: I've never seen this DataMirror thing before.  That looks useful for testing

File: src/test/scala/cookbook/RegOfVec.scala [hunk 15]
```diff
@@ -19,7 +19,7 @@ class RegOfVec extends CookbookTester(2) {
   //   Note that Seq.fill constructs 4 32-bit UInt literals with the value 0
   //   Vec(...) then constructs a Wire of these literals
   //   The Reg is then initialized to the value of the Wire (which gives it the same type)
-  val initRegOfVec = Reg(init = Vec(Seq.fill(4)(0.asUInt(32.W))))
+  val initRegOfVec = RegInit(Vec(Seq.fill(4)(0.asUInt(32.W))))
```
Comment (chick) [inline src/test/scala/cookbook/RegOfVec.scala]: Why use .asUInt here, rather than 0.U(...), also used below

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 7 above)
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: WireInit should be a separate PR. Will fix the literal format.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 8 above)
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Outside the scope of the Reg refactor... but should probably be part of a Wire refactor.

File: chiselFrontend/src/main/scala/chisel3/core/Reg.scala (hunk 9 above)
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Reg.scala]: Changing the behavior of Reg is next on the refactor list, because the behaviors right now are kind of confusing. This was probably the least-worst solution suggested during the meeting, though if you have a better solution we can discuss that.

Will change ordering, which will be consistent with RegNext.

File: src/main/scala/chisel3/compatibility.scala (hunk 10 above)
Comment (ducky64) [inline src/main/scala/chisel3/compatibility.scala]: Yeah, the main difference is that uses in chisel3._ should give a deprecated warning while uses in Chisel._ don't. Since the plan is to remove this old constructor from the chisel3._ package completely (and hopefully soon), I think this is fine.

File: src/main/scala/chisel3/util/Reg.scala (hunk 11 above)
Comment (ducky64) [inline src/main/scala/chisel3/util/Reg.scala]: RegEnable seemed to be more of a special case. Generally, I moved everything that could be built with the old Reg(t=..., next=..., init=...) constructor to core and nothing else.

File: src/test/scala/chiselTests/MultiClockSpec.scala (hunk 13 above)
Comment (ducky64) [inline src/test/scala/chiselTests/MultiClockSpec.scala]: Will fix.

File: src/test/scala/chiselTests/Reg.scala (hunk 14 above)
Comment (ducky64) [inline src/test/scala/chiselTests/Reg.scala]: Getting widths using the DataMirror reflection-like API or the old-style x.width might be something we should discuss.

File: src/test/scala/cookbook/RegOfVec.scala (hunk 15 above)
Comment (ducky64) [inline src/test/scala/cookbook/RegOfVec.scala]: Because that's how the last person wrote it?
I'll clean it up.

**Comment by ducky64** [comment]
Nits fixed.
